### PR TITLE
Features/ssao

### PR DESCRIPTION
**Pull request by dumitrurares14**
### Description

Added screen space ambient occlusion. We have a new graphics pipeline called ssao_pipeline that generates an AO texture that is sampled during the lighting pipeline to add on top of the final result.

Notes:

- There were some performance issues with this feature when you get too close to geometry, I tried to tackle it as best as possible without too much visual impact when you get too close to objects. Keep in mind that the effect might look different in those scenarios and is completely intended to avoid too much performance cost.
- SSAO is far from reality so don't expect to have realistic results. It looks nice, not realistic.
- For the moment this is a bit noisy; On top of the texture and other effects, it is not that noticeable and we can fix it by blurring it with the Gausian blur pipeline later. Don't complain about the noise, it's ok for the moment and some people even suggested it might even be nice for the artistic style..

BEFORE:
image

AFTER:
image

### Issues

https://bubonic-brotherhood.codecks.io/decks/8-graphics/card/1ak-ssao-user-story 
https://bubonic-brotherhood.codecks.io/card/1f8-ssao-shader 
https://bubonic-brotherhood.codecks.io/card/1f7-ssao-pipeline 

### Test criteria

- [ ] See if the output is something that we expect.
- [ ] Let me know if performance is impacted on different PCs. For me this works flawlessly
- [ ] Someone please test this on the steam deck.

### CODE REVIEW

File: engine/renderer/private/pipelines/ssao_pipeline.cpp [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+            randomFloats(generator) * 2.0 - 1.0,
+            0.0f,
+            0.0f); // padding
+        ssaoNoise.push_back(noise);
+    }
+
+    auto resources { _context->Resources() };
+    auto cmdBuffer = SingleTimeCommands(_context->VulkanContext());
+
+    bblog::warn("{} {} {}", ssaoKernel[0].x, ssaoKernel[0].y, ssaoKernel[0].z);
```
Comment (mmzala) [inline engine/renderer/private/pipelines/ssao_pipeline.cpp:148]: Delete unneeded log

File: engine/renderer/private/pipelines/ssao_pipeline.cpp [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    vk::PhysicalDeviceProperties properties {};
+    _context->VulkanContext()->PhysicalDevice().getProperties(&properties);
+
+    CreateBuffers();
+    CreateDescriptorSetLayouts();
+    CreateDescriptorSets();
+    CreatePipeline();
+}
+
+void SSAOPipeline::RecordCommands(vk::CommandBuffer commandBuffer, uint32_t currentFrame, MAYBE_UNUSED const RenderSceneDescription& scene) // NOLINT(*-include-cleaner)
```
Comment (mmzala) [inline engine/renderer/private/pipelines/ssao_pipeline.cpp:37]: `NOLINT(*-include-cleaner)` comment can be deleted

File: engine/renderer/private/pipelines/ssao_pipeline.cpp [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        .SetData(std::move(byteData))
+        .SetFlags(vk::ImageUsageFlagBits::eSampled)
+        .SetFormat(vk::Format::eR32G32B32A32Sfloat);
+
+    noiseImage.isHDR = true;
+
+    _ssaoNoise = _context->Resources()->ImageResourceManager().Create(noiseImage);
+    _pushConstants.ssaoNoiseIndex = _ssaoNoise.Index();
+
+    // sizeof(glm::vec4);
```
Comment (mmzala) [inline engine/renderer/private/pipelines/ssao_pipeline.cpp:188]: Delete unneeded comment

File: engine/renderer/private/pipelines/ssao_pipeline.cpp [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+        // scale magic to be more cluster near the actual fragment.
+        float scale = (float)i / 64.0;
+        scale = glm::mix(0.1f, 1.0f, scale * scale);
+        sample *= scale;
+        ssaoKernel.push_back(sample);
+    }
+
+    std::vector<glm::vec4> ssaoNoise;
+    for (unsigned int i = 0; i < 16; i++)
```
Comment (mmzala) [inline engine/renderer/private/pipelines/ssao_pipeline.cpp:135]: Use `uin32_t`

File: engine/renderer/private/pipelines/ssao_pipeline.cpp [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+        .SetDepthAttachmentFormat(vk::Format::eUndefined)
+        .BuildPipeline(_pipeline, _pipelineLayout);
+}
+void SSAOPipeline::CreateBuffers()
+{
+    // c++ side first
+    std::uniform_real_distribution<float> randomFloats(0.0, 1.0); // random floats between [0.0, 1.0]
+    std::default_random_engine generator;
+    std::vector<glm::vec4> ssaoKernel;
+    for (unsigned int i = 0; i < 64; ++i)
```
Comment (mmzala) [inline engine/renderer/private/pipelines/ssao_pipeline.cpp:117]: Use `uin32_t`

File: engine/renderer/public/pipelines/ssao_pipeline.hpp [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    vk::DescriptorSetLayout _descriptorSetLayout;
+    vk::DescriptorSet _descriptorSet;
+
+    std::shared_ptr<GraphicsContext> _context;
+    const GBuffers& _gBuffers;
+    const ResourceHandle<GPUImage> _ssaoTarget;
+
+    vk::PipelineLayout _pipelineLayout;
+    vk::Pipeline _pipeline;
+    const CameraResource& _camera;
```
Comment (mmzala) [inline engine/renderer/public/pipelines/ssao_pipeline.hpp:63]: Don't store the camera in the pipeline, you can get the descriptor set during the command recording, by getting the main camera through the GPUScene

File: engine/renderer/private/renderer.cpp [hunk 7]
```diff
@@ -114,13 +117,23 @@ Renderer::Renderer(ApplicationModule& application, Viewport& viewport, const std
         .SetDebugLabelColor(glm::vec3 { 0.0f, 1.0f, 1.0f })
         .AddOutput(_gBuffers->Shadow(), FrameGraphResourceType::eAttachment);
 
+    FrameGraphNodeCreation ssaoPass { *_ssaoPipeline };
+    ssaoPass.SetName("SSAO pass")
+        .SetDebugLabelColor(glm::vec3(0.87f))
+        .AddInput(_gBuffers->Attachments()[0], FrameGraphResourceType::eTexture)
+        .AddInput(_gBuffers->Attachments()[1], FrameGraphResourceType::eTexture)
+        .AddInput(_gBuffers->Attachments()[2], FrameGraphResourceType::eTexture)
+        .AddInput(_gBuffers->Attachments()[3], FrameGraphResourceType::eTexture)
```
Comment (mmzala) [inline engine/renderer/private/renderer.cpp:126]: The pass only uses position and normal buffers, so it should be specified as so here as well, otherwise the frame graph will think it has to put a memory barrier here for the other attachments here as well

File: engine/renderer/public/renderer.hpp [hunk 8]
```diff
@@ -48,6 +49,7 @@ class Renderer
     std::shared_ptr<GraphicsContext> GetContext() const { return _context; }
     DebugPipeline& GetDebugPipeline() const { return *_debugPipeline; }
     BloomSettings& GetBloomSettings() { return *_bloomSettings; }
+    SSAOPipeline& GetSSAOPipeline() const { return *_ssaoPipeline; }
```
Comment (mmzala) [inline engine/renderer/public/renderer.hpp:52]: SSAO Settings should be separate from the pass itself. There should be no need to get the pass outside of the renderer. Should be similar to something `BloomSettings`. But you can let this be for now, as we need to rework the renderer settings anyway.

File: engine/source/editor.cpp [hunk 9]
```diff
@@ -29,6 +29,8 @@
 #include <fstream>
 #include <imgui/misc/cpp/imgui_stdlib.h>
 
+#include <pipelines/ssao_pipeline.hpp>
```
Comment (mmzala) [inline engine/source/editor.cpp:32]: Should use `""`, instead of `<>`, as the latter is meant for system includes

File: game/engine_api.wren
Comment (mmzala) [inline game/engine_api.wren:1]: Should not be included in the PR

File: shaders/ssao.frag [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    Camera camera;
+};
+
+layout (set = 1, binding = 0) buffer SampleKernel { vec4 samples[]; } uSampleKernel;
+
+layout (location = 0) in vec2 texCoords;
+layout (location = 0) out vec4 outColor;
+
+
+const int kernelSize = 64;
```
Comment (mmzala) [inline shaders/ssao.frag:32]: Shouldn't this be a parameter that you pass to the shader?

File: engine/renderer/public/pipelines/lighting_pipeline.hpp [hunk 11]
```diff
@@ -43,4 +44,5 @@ class LightingPipeline final : public FrameGraphRenderPass
     vk::Pipeline _pipeline;
 
     const BloomSettings& _bloomSettings;
+    const ResourceHandle<GPUImage> _ssaoTarget;
```
Comment (mmzala) [inline engine/renderer/public/pipelines/lighting_pipeline.hpp:47]: No need to store this here, as it is only used to get the index to the push constants

File: engine/renderer/private/pipelines/ssao_pipeline.cpp [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+
+    commandBuffer.bindPipeline(vk::PipelineBindPoint::eGraphics, _pipeline);
+
+    commandBuffer.pushConstants<PushConstants>(_pipelineLayout, vk::ShaderStageFlagBits::eFragment, 0, _pushConstants);
+
+    commandBuffer.bindDescriptorSets(vk::PipelineBindPoint::eGraphics, _pipelineLayout, 0, { _context->BindlessSet() }, {});
+    commandBuffer.bindDescriptorSets(vk::PipelineBindPoint::eGraphics, _pipelineLayout, 1, { _descriptorSet }, {});
+    commandBuffer.bindDescriptorSets(vk::PipelineBindPoint::eGraphics, _pipelineLayout, 2, _camera.DescriptorSet(currentFrame), {});
+
+    // To add the descirpot sets for the kernel and noise buffers
```
Comment (mmzala) [inline engine/renderer/private/pipelines/ssao_pipeline.cpp:69]: Delete unneeded comment

File: engine/renderer/private/pipelines/ssao_pipeline.cpp [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+#include "gpu_scene.hpp"
+#include "graphics_context.hpp"
+#include "graphics_resources.hpp"
+#include "pipeline_builder.hpp"
+#include "resource_management/image_resource_manager.hpp"
+#include "shaders/shader_loader.hpp"
+#include "vulkan_context.hpp"
+
+#include <random>
+#include <resource_management/buffer_resource_manager.hpp>
```
Comment (mmzala) [inline engine/renderer/private/pipelines/ssao_pipeline.cpp:15]: Use `""` for our headers

File: engine/renderer/private/pipelines/ssao_pipeline.cpp [hunk 14]
```diff
@@ -0,0 +1,233 @@
+﻿#include "pipelines/ssao_pipeline.hpp"
+
+#include "../vulkan_helper.hpp"
```
Comment (mmzala) [inline engine/renderer/private/pipelines/ssao_pipeline.cpp:3]: Why go a directory back? You can just include it without doing so

**Review by mmzala — CHANGES_REQUESTED** [review]
Good work, I tested and it looks nice. I left some comments to clean up the code

File: engine/renderer/private/pipelines/ssao_pipeline.cpp (hunk 14 above)
Comment (dumitrurares14) [reply engine/renderer/private/pipelines/ssao_pipeline.cpp:3]: My bad, clion auto-complete probably incldued it like this

File: engine/renderer/private/pipelines/ssao_pipeline.cpp (hunk 1 above)
Comment (dumitrurares14) [reply engine/renderer/private/pipelines/ssao_pipeline.cpp:148]: Done 💪

File: shaders/ssao.frag (hunk 10 above)
Comment (dumitrurares14) [reply shaders/ssao.frag:32]: I chose to left it hard-coded 64, since the kernel is generated and cached on the CPU side with that size anyway and I don't want to regenerate it later.

File: engine/renderer/public/pipelines/lighting_pipeline.hpp (hunk 11 above)
Comment (dumitrurares14) [reply engine/renderer/public/pipelines/lighting_pipeline.hpp:48]: Removed 🫡

File: engine/renderer/private/renderer.cpp (hunk 7 above)
Comment (mmzala) [reply engine/renderer/private/renderer.cpp:128]: Make sure to remove targets 0 and 2 as they are not used here

File: engine/renderer/public/pipelines/ssao_pipeline.hpp (hunk 6 above)
Comment (mmzala) [reply engine/renderer/public/pipelines/ssao_pipeline.hpp:63]: Make sure to do this as well

File: game/engine_api.wren
Comment (mmzala) [inline game/engine_api.wren:1]: Remove this file from the PR

File: shaders/ssao.frag (hunk 10 above)
Comment (mmzala) [reply shaders/ssao.frag:32]: Alright, a nitpick but can you fix the formatting, it bothers me that the variable is directly above the main function XD

**Review by mmzala — COMMENTED** [review]
There are still a couple of things that need to be changed before merging, since some of them we don't want to have on main

**Review by mmzala — APPROVED** [review]
Nice work!
